### PR TITLE
Add `runWithoutSecurity` method

### LINKED ISSUE #61: Create `run` method without security manager
Whenever running `andy` locally, there is no need for the `AndySecurityManager` to take over control of the permissions, and thus this should be disabled when running locally. For this, a method `runWithoutSecurityManager` should be created which does not add the `AddSecurityManagerStep` to the execution flow, this way no security manager is added to the `andy` process.

### PR DESCRIPTION
**Pull request by thijsnulle**
This PR adds a `runWithoutSecurityManager` method which will be used in the `andy-maven-plugin` to ensure that whenever Andy is run locally, the security manager won't interfere with anything since it isn't necessary to check for students breaking Andy.

Closes #61

### CODE REVIEW

File: src/main/java/nl/tudelft/cse1110/andy/execution/ExecutionFlow.java
Comment (thijsnulle) [reply src/main/java/nl/tudelft/cse1110/andy/execution/ExecutionFlow.java]: Updated it to add both if the boolean `addSecurityManager` is true. Do you think renaming it something along the lines of `addSecurity` or `enableSecurity` would make more sense?

File: src/main/java/nl/tudelft/cse1110/andy/execution/ExecutionFlow.java
Comment (mauricioaniche) [inline src/main/java/nl/tudelft/cse1110/andy/execution/ExecutionFlow.java:18]: Can we have two methods here, buildWithSecurityChecks and build ?

Internally, we can use a Boolean to differentiate and avoid duplicate code, but for the clients of the API, i would avoid Boolean flags like this one.

File: src/main/java/nl/tudelft/cse1110/andy/execution/step/RunMetaTestsStep.java
Comment (mauricioaniche) [inline src/main/java/nl/tudelft/cse1110/andy/execution/step/RunMetaTestsStep.java:123]: Oh here, I guess the security check should also be false in case we are running it without security checks, right?

**Review by mauricioaniche — COMMENTED** [review]
Two comments

File: src/main/java/nl/tudelft/cse1110/andy/execution/ExecutionFlow.java
Comment (mauricioaniche) [reply src/main/java/nl/tudelft/cse1110/andy/execution/ExecutionFlow.java:18]: If that's too much work given the MetaStep, then it's ok for now.

File: src/main/java/nl/tudelft/cse1110/andy/execution/step/RunMetaTestsStep.java
Comment (mauricioaniche) [reply src/main/java/nl/tudelft/cse1110/andy/execution/step/RunMetaTestsStep.java:123]: Maybe the info about whether it's running in secure mode should go to the Context class, so that you know which mode to run in the meta test.

File: src/main/java/nl/tudelft/cse1110/andy/execution/ExecutionFlow.java
Comment (thijsnulle) [reply src/main/java/nl/tudelft/cse1110/andy/execution/ExecutionFlow.java]: Updated this to have two methods, `build` and `buildWithoutSecurityChecks`. I do not understand what is extra about the `MetaStep` but that might just be my misunderstanding of how the meta step works

File: src/main/java/nl/tudelft/cse1110/andy/execution/step/RunMetaTestsStep.java
Comment (thijsnulle) [reply src/main/java/nl/tudelft/cse1110/andy/execution/step/RunMetaTestsStep.java]: I think running it with security checks is a stronger assertion, and I think we want our tests working with the security manager. Also, I moved the boolean value to the `Context` class since this made indeed more sense.

Comment (mauricioaniche) [reply src/main/java/nl/tudelft/cse1110/andy/execution/step/RunMetaTestsStep.java]: If we run with the security mode off, the meta steps should also run _without_ the security steps. (With this current version of the code, even if Andy is run without security checks, it will do the security checks for the meta tests.)

What's happening here is that we replace the code with the mutated version, and run "Andy on top of Andy". That's why you see a new _ExecutionFlow_ being instantiated.

Comment (thijsnulle) [reply src/main/java/nl/tudelft/cse1110/andy/execution/step/RunMetaTestsStep.java]: I understand it now, I just assumed this was only run in our local tests but I get it. I changed it and built the `ExecutionFlow` based on the boolean value of the `Context` passed to the method.

**Review by martinmladenov — COMMENTED** [review]
Looks good. I have just one comment: in the method and field names, there are many references to "security manager". I think it would be better to change that to just "security" (i.e. `runWithoutSecurity`) or "security features", or something else that's more general than "security manager" since that also disables other security checks we perform (like the `SourceCodeSecurityCheckStep`, which has nothing to do with the SecurityManager, as well as other security steps we might add in the future).

**Comment by martinmladenov** [comment]
We should also modify the Maven plugin to run Andy without security features by default, since the security manager sometimes fails for no obvious reason depending on OS, JVM version, etc, and this could cause issues for students who want to run Andy locally. There should also be an optional argument to run it with security checks so that we can execute the security steps when we run the integration tests with the assignments repo.

**Review by martinmladenov — APPROVED** [review]

**Comment by thijsnulle** [comment]
@martinmladenov 
> We should also modify the Maven plugin to run Andy without security features by default, since the security manager sometimes fails for no obvious reason depending on OS, JVM version, etc, and this could cause issues for students who want to run Andy locally. There should also be an optional argument to run it with security checks so that we can execute the security steps when we run the integration tests with the assignments repo.

Yes, this is something that we will change, but for this implementation, the method needs to be added to Andy itself. Once this PR is merged, it will be updated in the andy-maven-plugin repository!

**Comment by thijsnulle** [comment]
@martinmladenov 
> Looks good. I have just one comment: in the method and field names, there are many references to "security manager". I think it would be better to change that to just "security" (i.e. `runWithoutSecurity`) or "security features", or something else that's more general than "security manager" since that also disables other security checks we perform (like the `SourceCodeSecurityCheckStep`, which has nothing to do with the SecurityManager, as well as other security steps we might add in the future).

I am with you on this one, do you think the method name `runWithoutSecurity` and the variable name `enableSecurity` in the `Context` class are good?

**Comment by martinmladenov** [comment]
> @martinmladenov
> 
> > Looks good. I have just one comment: in the method and field names, there are many references to "security manager". I think it would be better to change that to just "security" (i.e. `runWithoutSecurity`) or "security features", or something else that's more general than "security manager" since that also disables other security checks we perform (like the `SourceCodeSecurityCheckStep`, which has nothing to do with the SecurityManager, as well as other security steps we might add in the future).
> 
> I am with you on this one, do you think the method name `runWithoutSecurity` and the variable name `enableSecurity` in the `Context` class are good?

Yes, this sounds good to me!

**Comment by martinmladenov** [comment]
> @martinmladenov
> 
> > Looks good. I have just one comment: in the method and field names, there are many references to "security manager". I think it would be better to change that to just "security" (i.e. `runWithoutSecurity`) or "security features", or something else that's more general than "security manager" since that also disables other security checks we perform (like the `SourceCodeSecurityCheckStep`, which has nothing to do with the SecurityManager, as well as other security steps we might add in the future).
> 
> I am with you on this one, do you think the method name `runWithoutSecurity` and the variable name `enableSecurity` in the `Context` class are good?

@thijsnulle 
Yes, this sounds good to me!

**Comment by Arraying** [comment]
This still seems quite relevant to me. I can sit down and fix the merge conflicts, that's no problem. Shall we go ahead with a merge here guys?

**Comment by mauricioaniche** [comment]
The goal here was to make sure that our maven plugin doesn't use the security manager. After all, it is meant for people to run it locally. It's not a big thing, though, if we run it with the security manager now that it is working properly. I'll leave it up to you and Martin to decide that!

**Comment by martinmladenov** [comment]
The security manager has to be disabled when running via Maven as its behaviour can be inconsistent on different JVMs (this is a Java issue, it's not something we can fix). I'll resolve the merge conflicts since they seem to affect mainly code that I have written.

File: src/main/java/nl/tudelft/cse1110/andy/execution/metatest/library/LibraryMetaTest.java
Comment (Arraying) [inline src/main/java/nl/tudelft/cse1110/andy/execution/metatest/library/LibraryMetaTest.java:102]: Are these meta tests run when running Andy through Maven?

Comment (martinmladenov) [reply src/main/java/nl/tudelft/cse1110/andy/execution/metatest/library/LibraryMetaTest.java:102]: Yes, they are. Running it through Maven runs the same steps as running it normally.

 Meta tests are run in their own "sub-context" and we want to configure this context not to use security if the main context doesn't.

Comment (Arraying) [reply src/main/java/nl/tudelft/cse1110/andy/execution/metatest/library/LibraryMetaTest.java:102]: LGTM then!

**Comment by Arraying** [comment]
Let's merge this, if there's any issues we can fix them as we find them!

**Pull request merged**
